### ISSUE TITLE
Proposal: New Haddock doc format

### ISSUE THREAD

**Issue by m-renaud** [haskell/containers#486]
Hey, I'd like to get everyone's thoughts on the following formatting change to the Haddocks used throughout the containers package.

## Old

- Everything on one line (for the most part)
- Complexity the first thing in the docs (important, but not THAT important)
- Examples expanded by default

![fromlistwithold](https://user-images.githubusercontent.com/700646/34804516-e88e6570-f62d-11e7-9a70-582fa4a2f79b.png)

## New

- First line is a one sentence summary
- Followed by longer description with more details on args and caveats
- "See also" line
- "Complexity:" linke
- Examples (collapsed by default)

![fromlistwithnew](https://user-images.githubusercontent.com/700646/34804568-24b7c2ee-f62e-11e7-8f8a-31092abfb4eb.png)

**Comment by treeowl** [comment]
I'm not sure why this was assigned to me.... I urge you to write a mock-up
by changing several entries to your desired format. I think efficiency
should probably come before the section on gory details, but I'd really
have to see it both ways to form a clear opinion. I don't know how to get
things collapsed by default, but that sounds like a useful thing generally.

On Jan 10, 2018 8:46 PM, "Matt Renaud" <notifications@github.com> wrote:

Assigned #486 <https://github.com/haskell/containers/issues/486> to @treeowl
<https://github.com/treeowl>.

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/haskell/containers/issues/486#event-1418901761>, or mute
the thread

**Comment by m-renaud** [comment]
> I'm not sure why this was assigned to me

I assigned you because you had been making changes to the Sequence docs and had expressed interest in improving the haddocks.

> I don't know how to get things collapsed by default, but that sounds like a useful thing generally.

This is pretty easy, the following haddock markup will have examples collapsed by default:

```haskell
-- ==== __Examples__
--
-- Converting an empty map to a list yields an empty list.
--
-- >>> toList empty == []
```

> I'd really have to see it both ways to form a clear opinion.

As you wish :)

*Ignore the O(n) at the beginning of the `toList` summary*

Exhibit 1: Complexity after summary:
![listsectionnew](https://user-images.githubusercontent.com/700646/34808311-edf58286-f642-11e7-897f-fe0dbf2bd556.png)

Exhibit 2: Complexity at end:
![listsectionnewcomplexityatend](https://user-images.githubusercontent.com/700646/34808314-f796433e-f642-11e7-8479-2748a57b346b.png)

Exhibit 3: Complexity at the end (but above "see also"):
![listsectionseealsobelowcomplexity](https://user-images.githubusercontent.com/700646/34808370-5bbda21c-f643-11e7-9df0-34767fe109f1.png)

**Comment by treeowl** [comment]
Your example should also include operations with longer discussions of
complexity.

Real HTML results are much easier to evaluate than images. In fact, I
decree that images of text are prohibited on this bug tracker unless they
are unavoidable or nearly so, because accessibility on the internet is
really important.


On Jan 10, 2018 11:18 PM, "Matt Renaud" <notifications@github.com> wrote:

I'm not sure why this was assigned to me

I assigned you because you had been making changes to the Sequence docs and
had expressed interest in improving the haddocks.

I don't know how to get things collapsed by default, but that sounds like a
useful thing generally.

This is pretty easy, the following haddock markup will have examples
collapsed by default:

-- ==== __Examples__---- Converting an empty map to a list yields an
empty list.---- >>> toList empty == []

I'd really have to see it both ways to form a clear opinion.

As you wish :)

*Ignore the O(n) at the beginning of the toList summary*

Exhibit 1: Complexity after summary:
[image: listsectionnew]
<https://user-images.githubusercontent.com/700646/34808311-edf58286-f642-11e7-897f-fe0dbf2bd556.png>

Exhibit 2: Complexity at end:
[image: listsectionnewcomplexityatend]
<https://user-images.githubusercontent.com/700646/34808314-f796433e-f642-11e7-8479-2748a57b346b.png>

Exhibit 3: Complexity at the end (but above "see also"):
[image: listsectionseealsobelowcomplexity]
<https://user-images.githubusercontent.com/700646/34808370-5bbda21c-f643-11e7-9df0-34767fe109f1.png>

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/haskell/containers/issues/486#issuecomment-356820307>,
or mute the thread

**Comment by m-renaud** [comment]
> Your example should also include operations with longer discussions of complexity.

Do you have one in mind? I didn't see any longer discussions of complexity anywhere in the existing haddocks (although I may have missed it).

> Real HTML results are much easier to evaluate than images...accessibility on the internet is
really important.

Agreed, accessibility is very important. I'll convert a few more of the more documented functions and update the haddocks and send out links.

**Comment by treeowl** [comment]
Data.Sequence.chunksOf is a bit longer that way. I think that module has a
few others. I also wonder if there's a non-horrible way to make it easier
for users to get back from the big-O statement to the general explanation
of what the variables mean. The m and n are pretty much always obvious, but
Data.Int{Set, Map} have word size, and Data.Sequence has the index (for
relevant operations).

**Comment by m-renaud** [comment]
> Data.Sequence.chunksOf is a bit longer that way

Okay, I'll take a stab at that one.

> ...to get back from the big-O statement to the general explanation of what the variables mean

We could just say that on the same line (which is already done in some cases):

*Complexity: O(n)* where *n* is the number of entries in the map.

*Complexity: O(i)* where *i* is the breakpont index.

Also, are we in agreement that examples collapsed by default is strictly an improvement? If so, I think that could be a nice improvement.

**Comment by treeowl** [comment]
Probably. I'm not sure about situations where a single one-line example is
enough to understand everything.

**Comment by m-renaud** [comment]
Yeah, functions with a single one-line example can probably be left top-level.

**Comment by m-renaud** [comment]
Updated haddocks now live. You can find the section with list conversion examples [here](http://www.m-renaud.com/haddocks/containers-0.5.10.2/Data-Map-Strict.html#g:19) and the longer `chunksOf` example [here](http://www.m-renaud.com/haddocks/containers-0.5.10.2/Data-Sequence.html#v:chunksOf). Let me know what you think!

Also, it seems like haddock doesn't like complicated math expressions, it just doesn't render it.
